### DISCUSSION TITLE
Building ProTools AAX version of Ctrlr with Projucer

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#489] [Q&A] [answered]
Hi everybody, I was wondering if anyone already tried exporting Ctrlr as an AAX plugin with Projucer?

It seems that the steps are : 

1. Register to https://developer.avid.com/  (done)
2. Accept the agreement (done)
3. Download the AAX SDK (done)
4. Apply for a Product Certificate for public release (that could be problematic with licence type compatibility)
5. Request an ID for Avid/Pace/iLok Notarization of the AAX plugin.

Do you know if we need something specific to Ctrlr to build the AAX apart from linking to the SDK directory in Projucer?

Thanks

Damien

**Comment by damiensellier** [comment]
I could export the AAX without any problem from the Projucer, that wasn't difficult, the only hurdle was to build the AAX library in Xcode from the SDK files, I forgot to do it and had some errors while building Ctrlr. 

Now I need to test it in ProTools. I'll keep you posted.

**Comment by damiensellier** [comment]
So, it seems that AAX plugins that are not certified and code-signed via the AVID developer program are not loaded in ProTools.
unsigned AAX plugin can only be loaded in ProTools developer's version, not Release & Beta.
It's a bit useless in the current state then :(
Not sure if it's going to be possible to publicly distribute Ctrlr AAX as a "working/loading" plugin.  I'll dig deeper into that later this week. That would be nice instead of going the VST wrapper way.

**Comment by dobo365** [comment]
Hi Damien, any success with this one?
Wanted to try ProTools Intro (free) to test one of my panel.
For the moment, ProTools users can use Blue Cat Audio Patchwork that can allow VST/AU to be used as AAX in ProTools.
Thx!

**Reply by damiensellier** [reply]
Hi Dominique,
I built the AAX successfully. The process was smooth with Projucer. But ProTools does not load it because it needs to be certified through their dev program.
However it should run in their version for developpers for testing. 
To be a registered dev you have to pay to pass the "exam" under your registered company name and get to the 1st step of their licensing rabit hole. :'(
If we want a working AAX Ctrlr plugin that works for every protools version it has to be certified by Pace under the manufacturer name CTRLR..., so I'm trying to see with them how we can arrange something because of its collaborative open source nature. 
And I also need to see how it would be possible to get the approval for AAX limited instance plugin for every pannels (different manufacturers).
For the moment Blue Cat's Patchwork wrapper seems to be the best and only solution until further notice.

**Reply by damiensellier** [reply]
I wrote to avid for a follow up about this. I'll keep you posted.

**Comment by a819984059** [comment]
这是来自QQ邮箱的自动回复邮件。
 
您好，我已收到您的邮件。

**Comment by damiensellier** [comment]


**Reply by damiensellier** [reply]
As for AU in Logic not validating or Ableton Live not showing up, I think it's coming from Projucer option "Plugin Is MIDI Effect" ticked which disables Audio IN/OUT causing crashes. 

![398003416-4bb958a5-9a38-485e-9f14-4d42da879c4e](https://github.com/user-attachments/assets/be183dba-7221-4b55-8b89-d16e149be1e4)

**Reply by damiensellier** [reply]
UPDATE 05.06.2025  : Still crashing when inserting on an Instrument track on Pro Tools Dev Build even with MIDI Effect Plugin option disabled.

**Comment by damiensellier** [comment]
ISSUE open on CtrlrX github : https://github.com/damiensellier/CtrlrX/issues/54

**Comment by damiensellier** [comment, marked as answer]
CtrlrX is now supporting AAX plugins for ProTools, you need to build an intermediate plugin with your panel unique IDs though. And it also requires enrolling into the Avid developer program to get access to the SDK 2.8.1 and be able to codesign the exported plugin with PACE iLok. 

Exported instance, if not codesign via PACE tools won't load in ProTools commercial version (not Dev build).

---

### AAX Plugin-Specific Licensing & Avid SDK Terms

This plugin is designed to be compiled as an AAX plugin for Pro Tools. To do so, it utilizes the Avid AAX SDK (version 2.8.1 or newer).

The Avid AAX SDK is available under **commercial or open-source licensing**. For the purposes of this open-source project, we are using the AAX SDK under the terms of the **GNU General Public License v3 (GPLv3)**.

Due to the "copyleft" nature of the GPLv3, the entire distributed AAX plugin (including the portions of this project's code linked with the SDK) **must also be licensed under GPLv3**.

Therefore, any distributed AAX binaries, along with their corresponding source code, are subject to the terms of the GPLv3. You can find the full text of the GPLv3 [here](https://www.gnu.org/licenses/gpl-3.0.en.html).

**By building or using this plugin's AAX version, you acknowledge and agree to the terms of:**

* The [Avid AAX SDK License Agreement](https://developer.avid.com/aax)
* The [Avid Privacy Policy](https://www.avid.com/legal/privacy-policy-statement)

* AAX is a registered trademark of Avid Technologies, Inc.

### Crucial Note for Commercial Distribution of AAX Plugins (including those based on this project)

While this project is open-source under GPLv3, the Avid AAX SDK itself offers a dual-licensing model. **If you intend to commercially release a third-party plugin based on this project (or any AAX plugin derived from the Avid AAX SDK), you typically need to obtain a separate commercial license from Avid for the AAX SDK and comply with their specific commercial terms, which often include requirements for AAX plugin code signing via PACE Anti-Piracy tools.**

For information on commercial AAX development and licensing, please contact Avid directly at `audiosdk@avid.com` or visit the [Avid Developer Website](https://developer.avid.com/aax).

---

### AAX Support & Build Instructions

This project is built and tested with **JUCE 6.0.8**. To compile the AAX version of this plugin, you will also need to obtain the **Avid AAX SDK (version 2.8.1 or newer)** separately.

**To build the AAX version of this plugin:**

1.  **Obtain the Avid AAX SDK 2.8.1 (or newer compatible version):**
    * **Join the Avid Developer Program:** Visit the [Avid Developer website](https://developer.avid.com/aax/) and sign up for the AAX SDK program. This typically involves agreeing to their terms and creating an iLok account.
    * **Download the AAX SDK:** Once registered, navigate to the SDK download section and download "AAX SDK 2.8.1" (or the latest GPLv3-licensed version available for use with JUCE 6.0.8).
    * **Extract the SDK:** Extract the downloaded SDK archive to a stable, known location on your system (e.g., `C:\SDKs\AvidAAXSDK_2.8.1\` on Windows or `~/SDKs/AvidAAXSDK_2.8.1/` on macOS/Linux).

2.  **Build the AAX Library:**
    * Before building CtrlrX, you must compile the core AAX Library project included within the SDK. Navigate to the `Libs/AAXLibrary` directory within your extracted AAX SDK and build the project using your chosen IDE (e.g., Visual Studio on Windows, Xcode on macOS). This step generates the necessary static library (`.lib` or `.a`) that CtrlrX will link against.

3.  **Configure your build environment:**
    * **If using JUCE's Projucer (recommended for JUCE 6.0.8):**
        * Open the Projucer application (from your JUCE 6.0.8 clone).
        * Go to **"Settings"** (gear icon or File > Global Paths).
        * Under "SDK Locations" or "Module Paths", locate the entry for **"AAX SDK"** and set its path to the root directory of your extracted Avid AAX SDK (e.g., `C:\SDKs\AvidAAXSDK_2.8.1\`).
        * Open your `.jucer` project file, go to the "Exporters" section, and ensure the AAX format is enabled for your target platform.
    * **If using CMake:**
        * Your `CMakeLists.txt` will need to be configured to locate both your JUCE 6.0.8 root and the AAX SDK root.
        * You'll typically define variables (e.g., `JUCE_ROOT`, `AAX_SDK_ROOT`) and use `find_path` or explicit `set` commands, then pass `AAX_SDK_ROOT` to `juce_add_plugin` via the `AAX_SDK_PATH` argument. (Detailed CMake setup for JUCE 6.0.8 with external SDKs can be complex; a dedicated CMake guide might be needed for this if it's your primary build method).

4.  **Build the CtrlrX AAX binary:** Follow your project's general build instructions (e.g., via your IDE after Projucer generation, or directly with CMake) to compile the AAX target.
    * [Your specific build command/steps for AAX, e.g., for Projucer: "Open the .jucer project in Projucer, select the AAX target, and click 'Save and Open in IDE'. Then build in your IDE." or for CMake: `cmake --build . --config Release --target CtrlrX_AAX`]

**Additional Notes for Development and Distribution:**

* **Pro Tools Developer Version:** To test unsigned AAX plugins, you will likely need a special "Pro Tools Developer" version of Pro Tools, available through the [Avid Developer Program](https://developer.avid.com/aax).

* **PACE Codesigning:** For any distributed AAX plugin (commercial or free), you will generally need to go through the AVID authorization process via PACE codesigning tools. This is a requirement for plugins to load in standard (non-developer) Pro Tools versions. This process is managed by Avid and PACE.

---

![ezgif-2281c246002702](https://github.com/user-attachments/assets/c828aa84-d1da-4231-9f58-224ae50345f2)


ADDED on 2025.06.24 for CtrlrX v5.6.34 commit https://github.com/damiensellier/CtrlrX/commit/798c9295115bb4099840692473859ac127b42afe
